Strange WebSocket behavior since version 3.7

Close code of WebSocket became different after upgrading aiohttp from version 3.6.3 to version 3.7.2.
```
$ python -V
Python 3.9.0
```
server.py:
```python
from aiohttp import web

async def ws_handler(request):
    ws = web.WebSocketResponse()
    await ws.prepare(request)
    print('Closing')
    await ws.close()
    return ws

app = web.Application()
app.add_routes([web.get('/ws', ws_handler)])
web.run_app(app, port=8888)
```
Server logs are the same for all versions of aiohttp.

client.py:
```python
import asyncio
import aiohttp

async def main():
    async with aiohttp.ClientSession() as session:
        async with session.ws_connect('http://localhost:8888/ws') as ws:
            print('Sending')
            await ws.send_str('some data')
            await asyncio.sleep(1)
            print('Receiving')
            await ws.receive()
            print('Close code:', ws.close_code)

asyncio.run(main())
```
Client logs (**aiohttp==3.6.3**):
```
$ python client.py
Sending
Receiving
Close code: 1000
```
Client logs (**aiohttp==3.7.2**):
```
$ python client.py
Sending
Receiving
Close code: 1006
```

@bdraco Fancy another websocket one?

Issue here is the client receives a Close frame, then tries to send a Close frame in response, but then hits this error:
```
../hacking/aiohttp/aiohttp/client_ws.py:360: in receive
    await self.close()
../hacking/aiohttp/aiohttp/client_ws.py:264: in close
    await self._writer.close(code, message)
../hacking/aiohttp/aiohttp/http_websocket.py:718: in close
    await self._send_frame(
../hacking/aiohttp/aiohttp/http_websocket.py:657: in _send_frame
    self._write(header + mask + message)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <aiohttp.http_websocket.WebSocketWriter object at 0x7f4c20d17ca0>, data = b'\x88\x82h!\x9e\xdak\xc9'

    def _write(self, data: bytes) -> None:
        if self.transport.is_closing():
>           raise ConnectionResetError("Cannot write to closing transport")
E           ConnectionResetError: Cannot write to closing transport

../hacking/aiohttp/aiohttp/http_websocket.py:684: ConnectionResetError
```

This suggests the server has closed the connection prematurely. The server should wait for the Close frame from the client before closing the connection: https://datatracker.ietf.org/doc/html/rfc6455#section-1.4

--- a/tests/test_web_websocket_functional.py
+++ b/tests/test_web_websocket_functional.py
@@ -1152,9 +1152,30 @@ async def test_websocket_shutdown(aiohttp_client: AiohttpClient) -> None:
     assert websocket.closed is True
 
 
-@pytest.mark.xfail(reason="close never reaches client per issue #5180")
 async def test_ws_close_return_code(aiohttp_client: AiohttpClient) -> None:
     """Test that the close code is returned when the server closes the connection."""
+
+    async def handler(request: web.Request) -> web.WebSocketResponse:
+        ws = web.WebSocketResponse()
+        await ws.prepare(request)
+        await ws.receive()
+        await ws.close()
+        return ws
+
+    app = web.Application()
+    app.router.add_route("GET", "/", handler)
+    client = await aiohttp_client(app)
+    resp = await client.ws_connect("/")
+    await resp.send_str("some data")
+    msg = await resp.receive()
+    assert msg.type is aiohttp.WSMsgType.CLOSE
+    assert resp.close_code == WSCloseCode.OK
+
+
+async def test_abnormal_closure_when_server_does_not_receive(
+    aiohttp_client: AiohttpClient,
+) -> None:
+    """Test abnormal closure when the server closes and a message is pending."""
 
     async def handler(request: web.Request) -> web.WebSocketResponse:
         ws = web.WebSocketResponse()
@@ -1168,5 +1189,6 @@ async def test_ws_close_return_code(aiohttp_client: AiohttpClient) -> None:
     resp = await client.ws_connect("/")
     await resp.send_str("some data")
     await asyncio.sleep(0.1)
-    await resp.receive()
-    assert resp.close_code is WSCloseCode.OK
+    msg = await resp.receive()
+    assert msg.type is aiohttp.WSMsgType.CLOSE
+    assert resp.close_code == WSCloseCode.ABNORMAL_CLOSURE

